pandera/decorators.py

```diff
@@ -410,6 +413,34 @@ def _wrapper(
     return _wrapper
 
 
+@overload
+def check_types(
+    wrapped: F,
+    *,
+    head: Optional[int] = None,
+    tail: Optional[int] = None,
+    sample: Optional[int] = None,
+    random_state: Optional[int] = None,
+    lazy: bool = False,
+    inplace: bool = False,
+) -> F:
+    ...
+
+
+@overload
+def check_types(
+    wrapped: None = None,
+    *,
+    head: Optional[int] = None,
+    tail: Optional[int] = None,
+    sample: Optional[int] = None,
+    random_state: Optional[int] = None,
+    lazy: bool = False,
+    inplace: bool = False,
+) -> Callable[[F], F]:
+    ...
```

Comment:
@cristianmatache what is the purpose of this?

Reply:
@cosmicBboy not sure if you mean the second overload or the typing overall. I will answer both.

1. Why do we need the second overload? The first overload is matched when `@check_types` is used with no args (it takes a function ans returns a function with the same signature), and the second  overload is matched when the decorator is used with arguments like `@check_types(head=3)` (this returns a function/callable that takes in any function F and returns a function with the same signature as F.

2. Why do we need typing?
See the PR description:
Decorators were mostly resolving to Callable, this limits static type checking (e.g., with mypy) on the underlying function (for arguments that are not DataFrame-s for example). Moreover, this means that autocompletion in IDEs such as PyCharm does not work.

In other words, with this fix the decorator is now taking a function with some signature and returns a function with the same signature. While in the past it was returning just a Callable which is the same as Callable[..., Any] which means that the type checker was losing the signature of the underlying function. This also has an impact on IDEs like PyCharm because it kills the autocompletion feature. See the "before" and "after" sections for an actual example.

Reply:
cool, thanks for the explanation!